### ISSUE TITLE
permute() is slowed down massively in 1.5 compared to 1.4

### ISSUE THREAD

**Issue by pyscorcher** [pytorch/pytorch#37142]
## 🐛 Bug

Backpropagation through `tensor.permute()` is a lot slower in pytorch 1.5 compared to pytorch 1.4. 
I'm not exactly sure what is causing this, but it seems like a bug.(?)

## To Reproduce

Run following file with pytorch 1.4 and pytorch 1.5. (You might have to adjust the number of iterations or size of the input tensor if it takes too much or too little time)

```
import torch
import torch.nn as nn

class Perm(nn.Module):
    def __init__(self):
        super().__init__()
        self.s = nn.Parameter(torch.tensor([1.0]))

    def forward(self, x):
        y = x.mean(dim=(2, 3), keepdim=True) # = torch.nn.functional.adaptive_avg_pool2d(x, output_size=(1, 1))
        y = y.permute(0, 2, 3, 1)
        y = y*self.s  # without this, there is next to no difference
        y = y.permute(0, 3, 1, 2)
        y = x * y
        return y

device = torch.device('cuda:0') #cpu even worse difference
net = Perm().to(device)
opt = torch.optim.Adam(params=net.parameters(), lr=0.001)

def train(device=device, iters=10):
    x = torch.zeros((10, 64, 256, 256)).to(device)
    for k in range(iters):
        y = net(x).sum()
        opt.zero_grad()
        y.backward()
        opt.step()

if __name__ == '__main__':
    import timeit
    print(timeit.timeit("train()", setup="from __main__ import train", number=10))
```

## Expected behavior

I expected both versions to have a similar run time. However if th computations are done on the cpu (change the part in the code): 
* Runtime on cpu: v1.4: 5.07s / v1.5: 21.5s
* Runtime on gpu: v1.4: 0.665s/ v1.5: 3.31s

## Environment

 - PyTorch Version (e.g., 1.0): 1.4 and 1.5
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): - 
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: nvidia  P6000
 - Any other relevant information: -

**Comment by VitalyFedyunin** [comment]
Might be memory formats related, I will take a look.

**Comment by pyscorcher** [comment]
@VitalyFedyunin I suspected something like that - if you remove the line between the two permute() calls you don't have that slowdown. But I don't even know nearly enough to comment about that.

**Comment by lucmos** [comment]
I *think* I'm hitting the same problem!

I do not use directly `permute` but I use `einsum` that probably calls `permute` behind.

A small comparison of my train loop speed:

7 it/s in PyTorch 1.5
45 it/s in PyTorch 1.3

(I skipped the 1.4 since I am hitting [this](https://github.com/pytorch/pytorch/issues/32395))

**Comment by ngimel** [comment]
@lucmos do you have a repro script for your problem?

**Comment by lucmos** [comment]
@ngimel No, not at the moment, I'm sorry. 
It's a messy research code, and I only noticed a huge speed loss with PyTorch 1.5. I'm not sure what is the cause thus I'm not able to easily create a minimum working example.

**Comment by ngimel** [comment]
For the script above, the reason for the slowdown is that in the line `y = x * y` our nhwc propagation heuristics sees y in nhwc format and makes result nhwc too, which was not happening in 1.4. The immediate workaround for this is to modify line before that so that `y` is explicitly made regular contiguous:
```
y = y.permute(0, 3, 1, 2).to(memory_format=torch.contiguous_format)
```
It is still a pretty fast operation because it won't invoke any memory copy, `y` sizes are `(10,64,1,1)`, and it can be viewed as channels-last and regular contiguous without any memory movement. Slightly modified script (removes h2d copy in every iteration) is below, I'm getting 0.11 s on pytorch 1.4 and 0.13 s on master, which is still a regression and we'll look into it, but definitely not as dramatic. Before the fix, excluding h2d copy the results were 0.11/0.60. 
We'll fix the propagation heuristics so that nhwc propagation is not triggered in cases like this. Please note that for tensors with sizes XY11 it's impossible to determine whether they should be viewed as channels-last or not based on strides only, and our guess on whether user wants to propagate channels-last layout can be wrong. Hopefully, this problem is limited to only XY11 tensors.  
Also, to be clear, `.permute` operation itself is not to blame, `.permute` is a fast operation on tensor metadata, it does not invoke memcopies and it did not change between 1.4 and 1.5. The propagation of channels-last layout did change, and this is what can cause slowdowns. 
Modified benchmarking script below
```
import torch
import torch.nn as nn

class Perm(nn.Module):
    def __init__(self):
        super().__init__()
        self.s = nn.Parameter(torch.tensor([1.0]))

    def forward(self, x):
        # = torch.nn.functional.adaptive_avg_pool2d(x, output_size=(1, 1))
        y = x.mean(dim=(2, 3), keepdim=True)
        y = y.permute(0, 2, 3, 1)  # channels to last dim
        y = y*self.s  # without this, there is next to no difference
        # channels back to second dim
        y = y.permute(0, 3, 1, 2).to(memory_format=torch.contiguous_format)
        y = x * y
        return y

device = torch.device('cuda:0')  # can be changed to 'cpu'
net = Perm().to(device)
opt = torch.optim.Adam(params=net.parameters(), lr=0.001)

def train(device=device, iters=10):
    x = torch.zeros(10, 64, 256, 256, requires_grad=False, device=device)
    for k in range(iters):
        y = net(x).sum()
        opt.zero_grad()
        y.backward()
        opt.step()

if __name__ == '__main__':
    import timeit
    print(timeit.timeit("train()", setup="from __main__ import train", number=10))
```

**Comment by ezyang** [comment]
@ngimel what's the planned change to the propagation rules?

**Comment by VitalyFedyunin** [comment]
@ezyang implementing form of #32078 will remove ambiguity of NC11 cases. Besides that we are looking to add warning if operators are called with tensors in different memory formats.

**Comment by hengck23** [comment]
note that this also occurs for

```
   # very slow in 1.5/1.6 but ok for 1.4 !!!!
    input = input.transpose(0,1,4,2,3)  
    input = torch.from_numpy(input).float()
```


```
   # corrected version
    input = input.transpose(0,1,4,2,3).copy()  
    input = torch.from_numpy(input).float()
```

**Comment by ezyang** [comment]
@pyscorcher Hey, do you have the original, unreduced example, by any chance? I'm curious what the semantic content of the original script was supposed to be.

**Comment by gchanan** [comment]
We are moving this to 1.6.0 because this is the only report and we don't have a fix in yet.

**Comment by pyscorcher** [comment]
> @pyscorcher Hey, do you have the original, unreduced example, by any chance? I'm curious what the semantic content of the original script was supposed to be.

Sorry for the late reply, unfortunately I don't have the original code anymore, but it is not difficult to see: 

It was an implementation of a squeeze-and-excitation block of the SE-nets (see https://arxiv.org/pdf/1709.01507.pdf) The `permute()` was used to apply two `Linear()` maps channel-wise (on a `NxCxHxW = NxCx1x1` tensor, that is an image that was reduced by global average pooling).

But this code was obviously a little bit clumsy as you can easily replace the permute+Linear by a 2d-convolution with an 1x1-kernel.

**Comment by VitalyFedyunin** [comment]
Fixed for particular case with #37968 but requires permutations/tags/anything else as more permanent solution.
